### ISSUE TITLE
Performance of NPZ generation

### ISSUE THREAD

**Issue by napulen** [napulen/AugmentedNet#61]
The dataset generation needs to be more efficient.

Using `/usr/bin/time --verbose` to measure the memory usage and time elapsed of the application.

This is https://github.com/napulen/AugmentedNet/commit/52bdbe9933fbe48fb352e5601d2c68861ef796db
```
        Command being timed: "python -m AugmentedNet.dataset_npz_generator --collections abc"
        User time (seconds): 102.97
        System time (seconds): 2.65
        Percent of CPU this job got: 102%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:43.36
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1958924
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 464164
        Voluntary context switches: 59
        Involuntary context switches: 185
        Swaps: 0
        File system inputs: 0
        File system outputs: 27240
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

This is https://github.com/napulen/AugmentedNet/commit/dc743377c22c675d0b328c46dbd96d1f5cc33e12 It seems that `np.concatenate` is more expensive and time-consuming because it reallocates memory for every call, rather than appending.
```
Command being timed: "python -m AugmentedNet.dataset_npz_generator --collections abc"
        User time (seconds): 127.86
        System time (seconds): 12.11
        Percent of CPU this job got: 101%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:17.93
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2009264
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 775359
        Voluntary context switches: 52
        Involuntary context switches: 350
        Swaps: 0
        File system inputs: 8
        File system outputs: 27240
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

**Comment by napulen** [comment]
I borrowed the implementation of an "efficient" numpy array that grows dynamically in https://github.com/napulen/AugmentedNet/commit/3b4896c3e6135755b281101454d2bbc1ad8e2acb

```
        Command being timed: "python -m AugmentedNet.dataset_npz_generator --collections abc"
        User time (seconds): 129.30
        System time (seconds): 11.87
        Percent of CPU this job got: 101%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:19.06
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1975328
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 795013
        Voluntary context switches: 56
        Involuntary context switches: 287
        Swaps: 0
        File system inputs: 0
        File system outputs: 27240
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

**Comment by napulen** [comment]
Here is the performance after implementing a `memmap` array in https://github.com/napulen/AugmentedNet/commit/336392efb13bbbcdbf7f2ddbfcdde57db628a81e

```
        Command being timed: "python -m AugmentedNet.dataset_npz_generator --collections abc"
        User time (seconds): 96.62
        System time (seconds): 2.28
        Percent of CPU this job got: 102%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:36.65
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 720052
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 462
        Minor (reclaiming a frame) page faults: 184727
        Voluntary context switches: 61
        Involuntary context switches: 207
        Swaps: 0
        File system inputs: 0
        File system outputs: 8800
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

**Comment by napulen** [comment]
The last one is apparently faster and much lighter. Hopefully, that scales up accordingly for bigger datasets.

**Comment by napulen** [comment]
Fixed after #62
